curator-x-discovery/src/main/java/org/apache/curator/x/discovery/DiscoveryPathConstructor.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.curator.x.discovery;
+
+/**
+ * Constructs ZooKeeper paths to services for service discovering.
+ */
+public interface DiscoveryPathConstructor {
+    /**
+     * Return the path where all service names registered.
+     *
+     * @return the base path of all services
+     */
+    String getBasePath();
+
+    /**
+     * Return the path where all instances of the service registered.
+     *
+     * @param serviceName service name
+     * @return  path to service instances
+     */
+    String getPathForInstances(String serviceName);
+
+    /**
+     * Return the path where specified instance registered.
+     *
+     * @param serviceName service name
+     * @param instanceId instance id
+     * @return path to concrete instance
+     */
+    String getPathForInstance(String serviceName, String instanceId);
```

Comment:
Do we need this ? From javadocs, I concludes to "/base-path/{serviceName}/instance-dir/{instanceId}". Seems that `getPathForInstances` will give us directory for `instanceId` and `ServiceDiscoveryImpl::queryForInstances` depends on this. We need service directory and instance directory, that is all ?

Reply:
Given what the javadoc says:
1. `getPathForInstances`: Return the path where all instances of the service registered.
2. `getPathForInstance`: Return the path where specified instance registered.

The second can't be an extension point to return arbitrary path. Otherwise, it violates the first.

Also, I don't think this `getPathForInstance` will work. Assumes:

* `getBasePath()` returns `/base-path`.
* `getPathForInstances(serviceName)` returns `/base-path/{serviceName}/instances/`.

Is it possible for `getPathForInstance(serviceName, instanceId)` to return paths other than `/base-path/{serviceName}/instances/{instanceId}` ? If it is possible, how should `Collection<ServiceInstance<T>> queryForInstances(String name)` be implemented ? I expect the current implementation will break.

I think `getPathForInstance(serviceName, instanceId)` should behave same as `ZKPaths.makePath(getPathForInstances(serviceName), instanceId)` for this customization to function correctly. 

> And I'm OK to postpone this ticket unless anyone shows a real-world use case.

The jira already presented a good use case from my perspective. I can continue the rest if above make sense to you and in case you are not available to push forward this. I don't think we should/can expect response from original author given his/her inactivity.

Reply:
I go through the original ticket an think that it's the implementor's responsibility to implement the method correctly.

The original ticket argues that the default `ZKPaths.makePath` cannot satisfy some cases. And `ServiceDiscoveryImpl` delegates the related methods to the new PathConstructor:

```java
    @Override
    public Collection<String> queryForNames() throws Exception {
        List<String> names = client.getChildren().forPath(pathConstructor.getBasePath());
        return ImmutableList.copyOf(names);
    }
    String pathForName(String serviceName) {
        return pathConstructor.getPathForInstances(serviceName);
    }
    @VisibleForTesting
    String pathForInstance(String serviceName, String instanceId) {
        return pathConstructor.getPathForInstance(serviceName, instanceId);
    }
```

Reply:
@tisonkun I will open a pull request for discussion.